lib/opam_solve.ml

```diff
@@ -327,14 +327,18 @@ module Multi_dir_context :
                 Opam.Pp.package_name name Opam.Pp.version version);
           acc)
 
+  (** Candidates must be returned in decreasing preference order *)
   let candidates t name =
     let map =
       List.fold_left t ~init:OpamPackage.Version.Map.empty
         ~f:(fun acc dir_context ->
           let candidates = candidates dir_context name in
           merge_candidates ~name acc candidates)
     in
-    OpamPackage.Version.Map.bindings map
+    (* The bindings of the version map will be sorted by increasing version.
+       We reverse it so it's in decreasing version order for the solver to pick
+       the highest satisfying version. *)
+    List.rev (OpamPackage.Version.Map.bindings map)
```

Comment:
I looked up because I was confused by `bindings` to have an order (since I would expect either a random order or an order that is somehow dependent on insertion order), but indeed that's what OCaml's map does by default:

> Return the list of all bindings of the given map. The returned list is sorted in increasing order of keys with respect to the ordering `Ord.compare`, where `Ord` is the argument given to `Map.Make`.

I guess this is an artifact on how it is implemented but given this is documented I guess it is part of the implementation guarantees so it should be fine!